trie/transition.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2021 go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+	"github.com/ethereum/go-verkle"
+)
+
+type TransitionTrie struct {
+	overlay *VerkleTrie
+	base    *SecureTrie
+	storage bool
+}
+
+func NewTransitionTree(base *SecureTrie, overlay *VerkleTrie, st bool) *TransitionTrie {
+	return &TransitionTrie{
+		overlay: overlay,
+		base:    base,
+		storage: st,
+	}
+}
+
+func (t *TransitionTrie) Base() *SecureTrie {
+	return t.base
+}
+
+// TODO(gballet/jsign): consider removing this API.
+func (t *TransitionTrie) Overlay() *VerkleTrie {
+	return t.overlay
+}
+
+// GetKey returns the sha3 preimage of a hashed key that was previously used
+// to store a value.
+//
+// TODO(fjl): remove this when StateTrie is removed
+func (t *TransitionTrie) GetKey(key []byte) []byte {
+	if key := t.overlay.GetKey(key); key != nil {
+		return key
+	}
+	return t.base.GetKey(key)
+}
+
+// Get returns the value for key stored in the trie. The value bytes must
+// not be modified by the caller. If a node was not found in the database, a
+// trie.MissingNodeError is returned.
+func (t *TransitionTrie) GetStorage(addr common.Address, key []byte) ([]byte, error) {
+	val, err := t.overlay.GetStorage(addr, key)
+	if err != nil {
+		return nil, fmt.Errorf("get storage from overlay: %s", err)
+	}
+	if len(val) != 0 {
+		return val, nil
+	}
+	// TODO also insert value into overlay
+	return t.base.GetStorage(addr, key)
+}
+
+// GetAccount abstract an account read from the trie.
+func (t *TransitionTrie) GetAccount(address common.Address) (*types.StateAccount, error) {
+	data, err := t.overlay.GetAccount(address)
+	if err != nil {
+		return nil, err
+	}
+	if data != nil {
+		// 	// no pointer to db available, should the "transition" be done at the reader level instead?
+		// 	if t.overlay.db.HasStorageRootConversion(address) {
+		// 		data.Root = t.overlay.db.StorageRootConversion(address)
+		// 	}
```

Comment:
@rjl493456442 explainer here: because the account root isn't saved in the verkle tree (there is no such thing), the `data.Root` field would always be empty, which would confuse the layer. So I implemented this conversion table that, given an address, would give the historical root. That might no longer be necessary to do in the new model. But if it is, we'd have to think on how to implement that.